server/expression/in_subquery.go

```diff
@@ -0,0 +1,259 @@
+// Copyright 2024 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"fmt"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+	"github.com/dolthub/go-mysql-server/sql/types"
+	vitess "github.com/dolthub/vitess/go/vt/sqlparser"
+
+	"github.com/dolthub/doltgresql/server/functions/framework"
+	pgtypes "github.com/dolthub/doltgresql/server/types"
+)
+
+// InSubquery represents a VALUE IN (SELECT ...) expression.
+type InSubquery struct {
+	leftExpr  sql.Expression
+	rightExpr *plan.Subquery
+
+	// These variables are used so that we can resolve the comparison functions once and reuse them as we iterate over rows.
+	// These are assigned in WithChildren, so refer there for more information.
+	leftLiteral   *Literal
+	rightLiterals []*Literal
+	compFuncs     []*framework.CompiledFunction
+}
+
+var _ vitess.Injectable = (*InSubquery)(nil)
+var _ sql.Expression = (*InSubquery)(nil)
+var _ expression.BinaryExpression = (*InSubquery)(nil)
+
+// NewInSubquery returns a new *InSubquery.
+func NewInSubquery() *InSubquery {
+	return &InSubquery{}
+}
+
+// Children implements the sql.Expression interface.
+func (in *InSubquery) Children() []sql.Expression {
+	return []sql.Expression{in.leftExpr, in.rightExpr}
+}
+
+var nilKey, _ = sql.HashOf(sql.NewRow(nil))
+
+// Eval implements the sql.Expression interface.
+func (in *InSubquery) Eval(ctx *sql.Context, row sql.Row) (any, error) {
+	if len(in.compFuncs) == 0 {
+		return nil, fmt.Errorf("%T: cannot Eval as it has not been fully resolved", in)
+	}
+
+	left, err := in.leftExpr.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	// The NULL handling for IN expressions is tricky. According to
+	// https://dev.mysql.com/doc/refman/8.0/en/comparison-operators.html#operator_in:
+	// To comply with the SQL standard, IN() returns NULL not only if the expression on the left hand side is NULL, but
+	// also if no match is found in the list and one of the expressions in the list is NULL.
+	// However, there's a strange edge case. NULL IN (empty list) return 0, not NULL.
```

Comment:
This is about how MySQL works, but is this also how Postgres works? We should link to Postgres documentation and follow Postgres rules, otherwise we're just introducing future bugs.

Reply:
It appears to also be how postgres works:

```sql
test=# select null in (null);
 ?column?
----------

(1 row)

test=# select 1 in (null);
 ?column?
----------

(1 row)

test=# select 1 in (1, null);
 ?column?
----------
 t
(1 row)

test=# select 1 in (null, 1);
 ?column?
----------
 t
(1 row)

test=# select 1 in (0);
 ?column?
----------
 f
(1 row)

test=# select 0 in (null, 1);
 ?column?
----------
```

Which makes sense, it's part of the standard. I'll update the comment.